Content.Server/Hallucinations/HallucinationsSystem.cs

```diff
@@ -0,0 +1,150 @@
+    {
+        if (proto == null)
+            return false;
+        if (!_proto.TryIndex<HallucinationsPrototype>(proto, out var prototype))
+            return false;
+        if (!_status.TryAddStatusEffect<HallucinationsComponent>(target, key, time, refresh))
+            return false;
+
+        var hallucinations = _entityManager.GetComponent<HallucinationsComponent>(target);
+        hallucinations.Proto = prototype;
+        UpdatePreset(hallucinations);
+        hallucinations.CurChance = prototype.MinChance;
+
+        return true;
+    }
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<HallucinationsComponent, TransformComponent>();
+        while (query.MoveNext(out var uid, out var stat, out var xform))
+        {
+            if (_timing.CurTime < stat.NextSecond)
+                continue;
+            var rate = stat.SpawnRate;
+            stat.NextSecond = _timing.CurTime + TimeSpan.FromSeconds(rate);
+
+            if (stat.CurChance < stat.MaxChance && stat.CurChance + stat.IncreaseChance <= 1)
+                stat.CurChance = stat.CurChance + stat.IncreaseChance;
+
+            if (!_random.Prob(stat.CurChance))
+                continue;
+
+            stat.SpawnedCount = 0;
+
+            var range = stat.Range * 4;
+
+            // To be sure that entity will see right entities
+            UpdatePreset(stat);
+
+            // More people - worse hallucinations
+            foreach (var (ent, comp) in _lookup.GetEntitiesInRange<HumanoidAppearanceComponent>(xform.MapPosition, range))
+            {
+
+                var newCoords = Transform(ent).MapPosition.Offset(_random.NextVector2(stat.Range));
+
+                if (stat.SpawnedCount >= stat.MaxSpawns)
+                    continue;
+                stat.SpawnedCount = stat.SpawnedCount += 1;
+
+                var hallucination = Spawn(_random.Pick(stat.Spawns), newCoords);
+                EnsureComp<VisibilityComponent>(hallucination, out var visibility);
+                _visibilitySystem.SetLayer(hallucination, visibility, (int) stat.Layer, false);
+                _visibilitySystem.RefreshVisibility(hallucination, visibilityComponent: visibility);
+            }
+
+            // If there is no one... You are hallucinations source too
+            var uidnewCoords = Transform(uid).MapPosition.Offset(_random.NextVector2(stat.Range));
+            if (stat.SpawnedCount >= stat.MaxSpawns)
+                continue;
+            stat.SpawnedCount = stat.SpawnedCount += 1;
+
+            var uidhallucination = Spawn(_random.Pick(stat.Spawns), uidnewCoords);
+            EnsureComp<VisibilityComponent>(uidhallucination, out var uidvisibility);
+            _visibilitySystem.SetLayer(uidhallucination, uidvisibility, (int) stat.Layer, false);
+            _visibilitySystem.RefreshVisibility(uidhallucination, visibilityComponent: uidvisibility);
+
+        }
+    }
+
+}
```

Comment:
All of this needs to be clientside. Adding another visibility layer which will have a significant impact on PVS is absolutely not worth it.

EntityManager is also a dependency on systems you do not need to get it again.